### PR TITLE
Fix #2499

### LINKED ISSUE #2499: dune keeps no longer existing files from data_only_dirs in _build
I have an rule with a dep to a source_tree (let's say it's named spec). The alias generates a dune file based on the content of spec. It mostly does what I expect. Except when I run this alias after removing a file from spec. Because the content of spec in _build will still contain the file which was removed in the real spec. It seems to be because I marked the spec directory with data_only_dirs.

I created a small project to reproduce https://github.com/Khady/dune-data-only-dirs-cleanup

```
$ dune clean
$ cat from_spec_a.sexp 
6
$ ls -l spec_a/**/*.txt
-rw-r--r-- 1 louis louis 0 Aug  2 10:16 spec_a/d1/1.txt
-rw-r--r-- 1 louis louis 0 Aug  2 10:16 spec_a/d1/2.txt
-rw-r--r-- 1 louis louis 0 Aug  2 10:16 spec_a/d1/3.txt
-rw-r--r-- 1 louis louis 0 Aug  2 10:32 spec_a/d2/1.txt
-rw-r--r-- 1 louis louis 0 Aug  2 10:32 spec_a/d2/2.txt
-rw-r--r-- 1 louis louis 0 Aug  2 10:32 spec_a/d2/3.txt
$ dune build @gen_spec_a
$ tree _build/default/spec_a
_build/default/spec_a
├── d1
│   ├── 1.txt
│   ├── 2.txt
│   └── 3.txt
└── d2
    ├── 1.txt
    ├── 2.txt
    └── 3.txt

2 directories, 6 files
$ rm -rf spec_a/d2
$ dune build @gen_spec_a
$ ls -l spec_a/**/*.txt
-rw-r--r-- 1 louis louis 0 Aug  2 10:16 spec_a/d1/1.txt
-rw-r--r-- 1 louis louis 0 Aug  2 10:16 spec_a/d1/2.txt
-rw-r--r-- 1 louis louis 0 Aug  2 10:16 spec_a/d1/3.txt
$ tree _build/default/spec_a
_build/default/spec_a
├── d1
│   ├── 1.txt
│   ├── 2.txt
│   └── 3.txt
└── d2
    ├── 1.txt
    ├── 2.txt
    └── 3.txt

2 directories, 6 files
$ cat from_spec_a.sexp 
6
```

### PR DESCRIPTION
**Pull request by ghost**
Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

### CODE REVIEW

**Comment by diml** [comment]
There was an issue with the interpretation of `(source_tree x)`: for directories without files, we would load the rules/cleanup stale files and directories. This PR fixes that by forcing the build system to look at directories without files.

File: src/build.ml [hunk 1]
```diff
@@ -162,10 +162,35 @@ let of_result_map ?targets res ~f =
 let memoize name t =
   Memo { name; t; state = Unevaluated }
 
+(* This is to force the rules to be loaded for directories without
+   files when depending on [(source_tree x)]. Otherwise, we wouldn't
+   clean up stale directories in directories that contain no file. *)
+let depend_on_dir_without_files =
+  let pred = Predicate.create ~id:(lazy (String "false")) ~f:(fun _ -> false) in
+  fun dir ->
+    Paths_glob (File_selector.create ~dir pred) >>^ ignore
```
Comment (rgrinberg) [inline src/build.ml]: FYI, I've added `Build.ignore` to make this a bit safer. E.g:

```
... |> Build.ignore
```

**Review by rgrinberg — APPROVED** [review]
Looking good. Needs a CHANGES entry as well :)

File: src/build.ml (hunk 1 above)
Comment (diml) [reply src/build.ml]: Thanks for the info, I updated the code

**Comment by diml** [comment]
Oops, indeed. I added an entry

**Pull request merged**